### PR TITLE
Iai ols benchmark

### PR DESCRIPTION
**Pull request by oojo12**
This PR adds an iai benchmark for the OLS algorithm that is part of linfa-linear.

It has the following benchmark characteristics:

n = 1_000, 10_000, and 100_000
num_features = 7
values = Non unique (all are **f64:** 7)

If this is accepted, I will post the results in #228

This is also partly addresses #103

### CODE REVIEW

**Comment by oojo12** [comment]
That's not a problem, though iai might be feature complete at this time it seems to use Valgrind and cachegrind on the backend for the benchmarking. I'm also not sure excluding setup code is more important than the consistency provided by iai.

If you still prefer criterion I can change it to that by next week and report the benchmark results on the other issue.

**Comment by YuhanLiin** [comment]
Iai still has a bunch of open issues and PRs in its repo, so it's definitely not feature-complete. The author is still active so I'm not sure what the exact status of the project is. Setup code is not a big deal in longer benchmarks but could sway the percentage significantly in shorter benchmarks. Another concern with iai is whether it adds all executed cycles for multithread code, since that would make it seem as if everything was run sequentially.

I've asked the Zulip chat about this, so do you mind waiting on this for a week to see if anyone says anything? If I don't get a response then we go with Criterion.

**Comment by oojo12** [comment]
Heard back from the author via email and thought I'd drop it here. 

According to Brook:


"""
Yeah, it's more or less dead. I certainly haven't had the energy or time to maintain it properly for a while now. I don't expect that to change in the near future either.
"""

**Comment by YuhanLiin** [comment]
I think we should just go ahead and use Criterion.

**Comment by oojo12** [comment]
Gotcha will do tomorrow.

File: datasets/src/generate.rs [hunk 1]
```diff
@@ -66,21 +67,23 @@ fn make_blob(
...
     num_rows: usize,
     num_feats: usize,
-    num_targets: usize,
     feat_distr: X,
     target_distr: Y,
-) -> Dataset<f64, f64>
+) -> DatasetBase<
+    ArrayBase<OwnedRepr<f64>, Dim<[usize; 2]>>,
+    ArrayBase<OwnedRepr<f64>, Dim<[usize; 1]>>,
+>
```
Comment (YuhanLiin) [inline datasets/src/generate.rs:80]: You can put `Dataset<f64, f64, Ix1>` instead.

File: datasets/src/generate.rs [hunk 2]
```diff
@@ -66,21 +67,23 @@ fn make_blob(
 /// use linfa_datasets::generate::make_dataset;
 /// let feat_distr = Laplace::new(0.5, 5. ).unwrap();
 /// let target_distr = DiscreteUniform::new(0, 5).unwrap();
-/// make_dataset(5, 5, 2, feat_distr, target_distr);
+/// make_dataset(5, 5, feat_distr, target_distr);
 /// ```
 pub fn make_dataset<X, Y>(
     num_rows: usize,
     num_feats: usize,
-    num_targets: usize,
```
Comment (YuhanLiin) [inline datasets/src/generate.rs:74]: You're changing this to single-target data right?

File: algorithms/linfa-linear/benches/ols_bench.rs [hunk 3]
```diff
@@ -0,0 +1,28 @@
+use criterion::{criterion_group, criterion_main, BenchmarkId, Criterion};
+use linfa::traits::Fit;
+use linfa_datasets::generate::make_dataset;
+use linfa_linear::LinearRegression;
+use statrs::distribution::{DiscreteUniform, Laplace};
+
+fn perform_ols(num_rows: usize) {
+    let feat_distr = Laplace::new(0.5, 5.).unwrap();
+    let target_distr = DiscreteUniform::new(0, 5).unwrap();
+    let num_feats: usize = 5;
```
Comment (YuhanLiin) [inline algorithms/linfa-linear/benches/ols_bench.rs:10]: Can you also run the benchmark with different numbers of features (don't make it too large though).

File: algorithms/linfa-linear/benches/ols_bench.rs [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+    let dataset = make_dataset(num_rows, num_feats, feat_distr, target_distr);
```
Comment (YuhanLiin) [inline algorithms/linfa-linear/benches/ols_bench.rs:11]: Move `make_dataset` outside of this function so that it's not included in the benchmark times.

**Comment by YuhanLiin** [comment]
Can you also add a benchmark for the GLM algorithm?

File: datasets/src/generate.rs (hunk 1 above)
Comment (oojo12) [reply datasets/src/generate.rs:80]: will do.

File: datasets/src/generate.rs (hunk 2 above)
Comment (oojo12) [reply datasets/src/generate.rs:74]: yes I am.

File: algorithms/linfa-linear/benches/ols_bench.rs (hunk 3 above)
Comment (oojo12) [reply algorithms/linfa-linear/benches/ols_bench.rs:10]: Not a problem, something like 5, 10, 20 sound good? Just want to know what the standard should be.

**Comment by oojo12** [comment]
yep. Just to make sure we are on the same page, following the glm.rs example will be sufficient?

File: algorithms/linfa-linear/benches/ols_bench.rs (hunk 3 above)
Comment (YuhanLiin) [reply algorithms/linfa-linear/benches/ols_bench.rs:10]: Only really need 5 and 10

File: algorithms/linfa-linear/benches/ols_bench.rs (hunk 4 above)
Comment (oojo12) [reply algorithms/linfa-linear/benches/ols_bench.rs:11]: gotcha.

File: datasets/src/generate.rs (hunk 2 above)
Comment (YuhanLiin) [reply datasets/src/generate.rs:74]: Wouldn't it be more flexible to support multi-target generation? Single-target is just a special-case for multi-target. A target with dimensions `[n, 1]` can be converted into single-target.

**Comment by YuhanLiin** [comment]
Yes GLM example is good

File: datasets/src/generate.rs (hunk 2 above)
Comment (YuhanLiin) [reply datasets/src/generate.rs:74]: I suggest adding a dataset converter method for this in `linfa/src/dataset/impl_dataset.rs`

Comment (oojo12) [reply datasets/src/generate.rs:74]: I agree. At the time when I pushed the change I thought linfa algs only supported single_target datasets. What I can do is leave it as a multi-target generation and make the suggested converter method in another PR. 

The question is would you rather I finish this work first or the dataset converter method?

Comment (YuhanLiin) [reply datasets/src/generate.rs:74]: Dataset converter, which this work depends on.

**Review by YuhanLiin — APPROVED** [review]

**Pull request merged**